src/ecc.rs

```diff
@@ -437,8 +457,104 @@ impl Mechanism for EccMechanism {
     }
 }
 
+fn kdf_type_to_hash_mech(mech: CK_EC_KDF_TYPE) -> KResult<CK_MECHANISM_TYPE> {
+    match mech {
+        CKD_SHA1_KDF => Ok(CKM_SHA_1),
+        CKD_SHA224_KDF => Ok(CKM_SHA224),
+        CKD_SHA256_KDF => Ok(CKM_SHA256),
+        CKD_SHA384_KDF => Ok(CKM_SHA384),
+        CKD_SHA512_KDF => Ok(CKM_SHA512),
+        CKD_SHA3_224_KDF => Ok(CKM_SHA3_224),
+        CKD_SHA3_256_KDF => Ok(CKM_SHA3_256),
+        CKD_SHA3_384_KDF => Ok(CKM_SHA3_384),
+        CKD_SHA3_512_KDF => Ok(CKM_SHA3_512),
+        _ => return err_rv!(CKR_MECHANISM_PARAM_INVALID),
+    }
+}
+
+#[derive(Debug)]
+struct ECDHOperation<'a> {
+    mech: CK_MECHANISM_TYPE,
+    kdf: CK_EC_KDF_TYPE,
+    hash: Option<CK_MECHANISM_TYPE>,
+    public: &'a [u8],
+    shared: &'a [u8],
+    finalized: bool,
+}
+
+impl ECDHOperation<'_> {
+    fn new_mechanism() -> Box<dyn Mechanism> {
+        Box::new(EccMechanism {
+            info: CK_MECHANISM_INFO {
+                ulMinKeySize: MIN_EC_SIZE_BITS as CK_ULONG,
+                ulMaxKeySize: MAX_EC_SIZE_BITS as CK_ULONG,
+                flags: CKF_DERIVE,
+            },
+        })
+    }
+
+    fn register_mechanisms(mechs: &mut Mechanisms) {
+        for ckm in &[CKM_ECDH1_DERIVE, CKM_ECDH1_COFACTOR_DERIVE] {
+            mechs.add_mechanism(*ckm, Self::new_mechanism());
+        }
+    }
+
+    fn derive_new<'a>(
+        mechanism: CK_MECHANISM_TYPE,
+        params: *const CK_ECDH1_DERIVE_PARAMS,
+    ) -> KResult<ECDHOperation<'a>> {
+        let p = unsafe { *params };
+        if p.kdf == CKD_NULL {
+            if p.pSharedData != std::ptr::null_mut() || p.ulSharedDataLen != 0 {
+                return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+            }
+        }
+        if p.pPublicData == std::ptr::null_mut() || p.ulPublicDataLen == 0 {
+            return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+        }
+
+        let shared = if p.pSharedData != std::ptr::null_mut()
+            && p.ulSharedDataLen != 0
+        {
+            unsafe {
```

Comment:
I think I was not satisfied with the `bytes_to_slice` not checking for the NULL pointer (only the len). This case is indeed invalid, but I thought it would be better to report error than crash.

I think the `bytes_to_vec()` can not handle this case either so I went to implement it from scratch. But I agree that it might be better to teach the `bytes_to_vec()` to handle empty pairs and safely return empty vec instead of checking the length like in the `aes.c` before every call that is non-mandatory value:

https://github.com/latchset/kryoptic/blob/eee73043e7232714f17c857e034e1bb65de8acaa/src/ossl/aes.rs#L246-L249

Reply:
yes, or we can have a mcro that takes an other parameter, to which we can key off different behaviors:
for example if you pass in an error that is the error to return if either of the parameters is null/0 instead of returning an empty vec.

Reply:
by which I mean you would add another implementation of the same macro when there is an additional parameter, I am ok for the default to return an empty vec I think.

Reply:
Changed to vec, fixed the `bytes_to_vec!()` macro and the obvious usages in aes. Skipped the more complicated/mandatory in kdf.